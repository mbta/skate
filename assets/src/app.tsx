// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
