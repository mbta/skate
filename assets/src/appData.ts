const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
