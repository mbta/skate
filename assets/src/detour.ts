import { LatLngLiteral } from "leaflet"
import { Shape, ShapePoint } from "./schedule"

export interface DetourShape {
  coordinates: ShapePoint[]
  directions: {
    instruction: string
  }[]
}

export interface OriginalRoute {
  routeName: string
  routeDescription: string
  routeOrigin: string
  routeDirection: string
  shape: Shape
  center: LatLngLiteral
  zoom: number
}
