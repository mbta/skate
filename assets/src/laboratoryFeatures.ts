import appData from "./appData"

const featureIsEnabled = (key: string): boolean => {
  const data = appData()

  if (data === undefined || data.laboratoryFeatures === undefined) {
    return false
  }

  const features = JSON.parse(data.laboratoryFeatures)
  if (features[key] === undefined) {
    return false
  }

  return features[key]
}

export default featureIsEnabled
