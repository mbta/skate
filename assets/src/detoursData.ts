import { array, Infer, number, string, type } from "superstruct"
import { DraftDetours } from "./models/detours"

export const DraftDetoursData = array(
  type({
    route: string(),
    direction: string(),
    name: string(),
    intersection: string(),
    active_since: number(),
  })
)
export type DraftDetoursData = Infer<typeof DraftDetoursData>

export const draftDetoursFromData = (
  draftDetoursData: DraftDetoursData
): DraftDetours =>
  draftDetoursData.map((draft) => ({
    route: draft.route,
    direction: draft.direction,
    name: draft.name,
    intersection: draft.intersection,
    activeSince: draft.active_since,
  }))
