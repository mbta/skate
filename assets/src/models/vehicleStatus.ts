import featureIsEnabled from "../laboratoryFeatures"
import { Vehicle } from "../realtime.d"

/** Where a vehicle is relative to its schedule.
 * Swiftly assigns this to all vehicles.
 */
export type OnTimeStatus = "on-time" | "early" | "late"

export enum HeadwaySpacing {
  VeryBunched = "very-bunched",
  Bunched = "bunched",
  Ok = "ok",
  Gapped = "gapped",
  VeryGapped = "very-gapped",
}

/** How the vehicle triangle should be drawn,
 * taking into account the vehicle's state and the settings
 */
export type DrawnStatus = OnTimeStatus | "off-course" | "ghost" | "plain"

export const onTimeStatus = (scheduleAdherenceSecs: number): OnTimeStatus => {
  const oneMinuteInSeconds = 60
  const sixMinutesInSeconds = 360

  if (scheduleAdherenceSecs < -oneMinuteInSeconds) {
    return "early"
  } else if (scheduleAdherenceSecs > sixMinutesInSeconds) {
    return "late"
  } else {
    return "on-time"
  }
}

export const drawnStatus = (vehicle: Vehicle): DrawnStatus => {
  if (vehicle.isOffCourse) {
    return "off-course"
  } else if (
    isShuttle(vehicle) ||
    (featureIsEnabled("headway_ladder_colors") &&
      vehicle.headwaySpacing !== null)
  ) {
    // Headway lines give the status instead of the vehicles
    return "plain"
  } else {
    return onTimeStatus(vehicle.scheduleAdherenceSecs)
  }
}

export const humanReadableScheduleAdherence = (vehicle: Vehicle): string =>
  vehicle.isOffCourse
    ? "Invalid"
    : humanReadableOnTimeStatus(onTimeStatus(vehicle.scheduleAdherenceSecs))

const humanReadableOnTimeStatus = (status: OnTimeStatus): string => {
  switch (status) {
    case "early":
      return "early"

    case "on-time":
      return "on time"

    case "late":
      return "late"
  }
}

export const humanReadableHeadwaySpacing = (
  spacing: HeadwaySpacing | null
): string => {
  switch (spacing) {
    case null:
      return "good"

    case HeadwaySpacing.VeryGapped:
      return "very gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "good"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very bunched"
  }
}

export const headwaySpacingToString = (spacing: HeadwaySpacing): string => {
  switch (spacing) {
    case HeadwaySpacing.VeryGapped:
      return "very-gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "ok"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very-bunched"
  }
}

export const statusClass = (status: DrawnStatus): string => {
  switch (status) {
    case "plain":
      return ""

    default:
      return status
  }
}

const isShuttle = (vehicle: Vehicle): boolean =>
  (vehicle.runId || "").startsWith("999")
