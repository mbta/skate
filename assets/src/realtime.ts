import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
