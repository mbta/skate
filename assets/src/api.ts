import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
