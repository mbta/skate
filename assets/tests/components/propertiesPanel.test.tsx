import React from "react"
import renderer from "react-test-renderer"
import PropertiesPanel, {
  handleSwipe,
} from "../../src/components/propertiesPanel"
import { HeadwaySpacing } from "../../src/models/vehicleStatus"
import { Ghost, Vehicle } from "../../src/realtime"
import { Route } from "../../src/schedule"

jest.spyOn(Date, "now").mockImplementation(() => 234000)

const route: Route = {
  id: "39",
  directionNames: {
    0: "Outbound",
    1: "Inbound",
  },
  name: "39",
}

describe("PropertiesPanel", () => {
  test("renders a vehicle", () => {
    const vehicle: Vehicle = {
      id: "v1",
      label: "v1-label",
      runId: "run-1",
      timestamp: 123,
      latitude: 0,
      longitude: 0,
      directionId: 0,
      routeId: "39",
      tripId: "t1",
      headsign: "Forest Hills",
      viaVariant: "X",
      operatorId: "op1",
      operatorName: "SMITH",
      bearing: 33,
      blockId: "block-1",
      headwaySecs: 859.1,
      headwaySpacing: HeadwaySpacing.Ok,
      previousVehicleId: "v2",
      scheduleAdherenceSecs: 0,
      scheduledHeadwaySecs: 120,
      isOffCourse: false,
      layoverDepartureTime: null,
      blockIsActive: false,
      dataDiscrepancies: [
        {
          attribute: "trip_id",
          sources: [
            {
              id: "swiftly",
              value: "swiftly-trip-id",
            },
            {
              id: "busloc",
              value: "busloc-trip-id",
            },
          ],
        },
      ],
      stopStatus: {
        stopId: "s1",
        stopName: "Stop Name",
      },
      timepointStatus: {
        fractionUntilTimepoint: 0.5,
        timepointId: "tp1",
      },
      scheduledLocation: null,
      routeStatus: "on_route",
    }

    const tree = renderer
      .create(
        <PropertiesPanel selectedVehicleOrGhost={vehicle} route={route} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a ghost", () => {
    const ghost: Ghost = {
      id: "ghost-trip",
      directionId: 0,
      routeId: "39",
      tripId: "trip",
      headsign: "headsign",
      blockId: "block",
      runId: "123-0123",
      viaVariant: "X",
      scheduledTimepointStatus: {
        timepointId: "t0",
        fractionUntilTimepoint: 0.0,
      },
    }

    const tree = renderer
      .create(<PropertiesPanel selectedVehicleOrGhost={ghost} route={route} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe("handleSwipe", () => {
  test("hides the panel on a right swipe", () => {
    const hidePanelCB = jest.fn()

    handleSwipe(hidePanelCB)("Right", null)
    expect(hidePanelCB).toHaveBeenCalled()
  })

  test("does not hide panel on other swipes", () => {
    const hidePanelCB = jest.fn()
    const handler = handleSwipe(hidePanelCB)
    handler("Left", null)
    handler("Up", null)
    handler("Down", null)

    expect(hidePanelCB).not.toHaveBeenCalled()
  })

  test("does not hide panel when map is swiped right", () => {
    const hidePanelCB = jest.fn()
    const map = document.createElement("div")
    map.id = "id-vehicle-map"
    handleSwipe(hidePanelCB)("Right", map)
  })
})
