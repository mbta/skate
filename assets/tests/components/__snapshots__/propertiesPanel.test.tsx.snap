// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertiesPanel renders a ghost 1`] = `
Array [
  <div
    className="m-properties-panel"
    id="m-properties-panel"
  >
    <div
      className="m-ghost-properties-panel"
    >
      <div
        className="m-properties-panel__header-wrapper"
      >
        <div
          className="m-properties-panel__header"
        >
          <div
            className="m-properties-panel__label"
          >
            <svg
              style={
                Object {
                  "height": 64,
                  "width": 64,
                }
              }
              viewBox="-32 -20 64 64"
            >
              <g
                className="m-vehicle-icon m-vehicle-icon--large ghost early-red"
              >
                <rect
                  className="m-vehicle-icon__label-background"
                  height={24}
                  rx={12}
                  ry={12}
                  width={64}
                  x={-32}
                  y={19}
                />
                <text
                  className="m-vehicle-icon__label m-vehicle-icon__label--normal"
                  dominantBaseline="central"
                  textAnchor="middle"
                  x="0"
                  y={31}
                >
                  0123
                </text>
                <g
                  transform="scale(0.7) translate(-24,-23)"
                >
                  <path
                    className="m-vehicle-icon__ghost-highlight"
                    d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
                    stroke-join="round"
                  />
                  <path
                    className="m-vehicle-icon__ghost-body"
                    d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
                    stroke-join="round"
                  />
                </g>
                <text
                  className="m-vehicle-icon__variant"
                  dominantBaseline="alphabetic"
                  textAnchor="middle"
                  x={0}
                  y={8}
                >
                  X
                </text>
              </g>
            </svg>
          </div>
          <div
            className="m-properties-panel__variant"
          >
            <div
              className="m-properties-panel__inbound-outbound"
            >
              Outbound
            </div>
            <div
              className="m-route-variant-name"
            >
              <span
                className="m-route-variant-name__route-id"
              >
                39_X
              </span>
               headsign
            </div>
          </div>
          <div
            className="m-properties-panel__close-ping"
          >
            <button
              className="m-close-button"
              onClick={[Function]}
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
          </div>
        </div>
        <ul
          className="m-tabs__tab-list"
        >
          <li
            className="m-tabs__tab m-tabs__tab--selected"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Status
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Run
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Block
          </li>
        </ul>
      </div>
      <div
        className="m-tabs__tab-panel"
      >
        <div
          className="m-ghost-properties-panel__no-waiver-banner"
        >
          <div
            className="m-ghost-properties-panel__no-waiver-banner-header"
          >
            <span
              className="m-ghost-properties-panel__no-waiver-banner-alert-icon"
            >
              <svg
                className="c-icon-alert-circle"
                viewBox="-3 -3 54 54"
              >
                <g
                  className="c-icon-alert-circle--highlighted"
                >
                  <circle
                    className="c-icon-alert-circle__outline"
                    cx="24"
                    cy="24"
                    r="27"
                  />
                  <circle
                    className="c-icon-alert-circle__fill"
                    cx="24"
                    cy="24"
                    r="22.59"
                  />
                  <g
                    className="c-icon-alert-circle__exclamation-point"
                  >
                    <path
                      d="m20.39 4.42h7.22a1.81 1.81 0 0 1 1.49 2.11l-1.49 23.22a1.73 1.73 0 0 1 -1.49 1.78h-4.24a1.73 1.73 0 0 1 -1.49-1.78l-1.49-23.22c-.07-1.13.61-2.11 1.49-2.11"
                    />
                    <path
                      d="m21.57 34.54h4.87a1.88 1.88 0 0 1 1.82 1.93v5.17a1.89 1.89 0 0 1 -1.82 1.94h-4.87a1.89 1.89 0 0 1 -1.83-1.94v-5.17a1.88 1.88 0 0 1 1.83-1.93"
                    />
                  </g>
                </g>
              </svg>
            </span>
            <div
              className="m-ghost-properties-panel__no-waiver-banner-title"
            >
              Unknown Ghost Bus - No Dispatcher Note
            </div>
          </div>
          A ghost bus or dropped trip has been automatically detected on this route. Please follow up with dispatch as needed, if a Dispatcher Note is generated - it will appear here.
        </div>
        <div
          className="m-properties-list"
        >
          <table
            className="m-properties-list__table"
          >
            <tbody>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Run
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  123-0123
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="m-properties-panel__modal-overlay"
    onClick={[Function]}
  />,
]
`;

exports[`PropertiesPanel renders a vehicle 1`] = `
Array [
  <div
    className="m-properties-panel"
    id="m-properties-panel"
  >
    <div
      className="m-vehicle-properties-panel"
    >
      <div
        className="m-properties-panel__header-wrapper"
      >
        <div
          className="m-properties-panel__header"
        >
          <div
            className="m-properties-panel__label"
          >
            <svg
              style={
                Object {
                  "height": 64,
                  "width": 64,
                }
              }
              viewBox="-32 -20 64 64"
            >
              <g
                className="m-vehicle-icon m-vehicle-icon--large on-time early-red"
              >
                <rect
                  className="m-vehicle-icon__label-background"
                  height={24}
                  rx={12}
                  ry={12}
                  width={64}
                  x={-32}
                  y={19}
                />
                <text
                  className="m-vehicle-icon__label m-vehicle-icon__label--normal"
                  dominantBaseline="central"
                  textAnchor="middle"
                  x="0"
                  y={31}
                >
                  1
                </text>
                <path
                  className="m-vehicle-icon__triangle"
                  d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
                  transform="scale(1) rotate(0) translate(-24,-22)"
                />
                <text
                  className="m-vehicle-icon__variant"
                  dominantBaseline="alphabetic"
                  textAnchor="middle"
                  x={0}
                  y={14}
                >
                  X
                </text>
              </g>
            </svg>
          </div>
          <div
            className="m-properties-panel__variant"
          >
            <div
              className="m-properties-panel__inbound-outbound"
            >
              Outbound
            </div>
            <div
              className="m-route-variant-name"
            >
              <span
                className="m-route-variant-name__route-id"
              >
                39_X
              </span>
               Forest Hills
            </div>
            <div
              className="m-properties-panel__schedule-adherence on-time early-red"
            >
              <div
                className="m-properties-panel__schedule-adherence-status-icon"
              >
                <svg
                  height="10"
                  width="10"
                >
                  <circle
                    cx="5"
                    cy="5"
                    r="5"
                  />
                </svg>
              </div>
              <div
                className="m-properties-panel__schedule-adherence-status-string"
              >
                on time
              </div>
              <div
                className="m-properties-panel__schedule-adherence-label"
              >
                (0 min early)
              </div>
            </div>
          </div>
          <div
            className="m-properties-panel__close-ping"
          >
            <button
              className="m-close-button"
              onClick={[Function]}
            >
              <span
                className=""
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "SVG",
                  }
                }
              />
            </button>
            <div
              className="m-properties-panel__last-gps-ping"
            >
              111s ago
            </div>
          </div>
        </div>
        <ul
          className="m-tabs__tab-list"
        >
          <li
            className="m-tabs__tab m-tabs__tab--selected"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Status
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Run
          </li>
          <li
            className="m-tabs__tab"
            onClick={[Function]}
          >
            <svg
              className="m-tabs__tab-status-icon"
              height="14"
              width="14"
            >
              <circle
                className="m-tabs__tab-status-icon-outer-circle"
                cx="7"
                cy="7"
                r="6"
              />
              <circle
                className="m-tabs__tab-status-icon-inner-circle"
                cx="7"
                cy="7"
                r="4"
              />
            </svg>
            Block
          </li>
        </ul>
      </div>
      <div
        className="m-tabs__tab-panel"
      >
        <div
          className="m-properties-list"
        >
          <table
            className="m-properties-list__table"
          >
            <tbody>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Run
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  run-1
                </td>
              </tr>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Vehicle
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  v1-label
                </td>
              </tr>
              <tr
                className="m-properties-list__property "
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Operator
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  PATTI SMITH #op1
                </td>
              </tr>
              <tr
                className="m-properties-list__property m-properties-list__property--last-login"
              >
                <td
                  className="m-properties-list__property-label"
                >
                  Last Login
                </td>
                <td
                  className="m-properties-list__property-value"
                >
                  4 hr 3 min; 1:38 PM
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          className="m-vehicle-properties-panel__location"
        >
          <div
            className="m-vehicle-properties-panel__latlng"
          >
            <div
              className="m-vehicle-properties-panel__label"
            >
              Current Location
            </div>
            <a
              className="m-vehicle-properties-panel__link"
              href="https://www.google.com/maps/dir/?api=1&destination=0,0&travelmode=driving"
              target="_blank"
            >
              Directions
            </a>
          </div>
          <div
            className="m-vehicle-properties-panel__next-stop"
          >
            <div
              className="m-vehicle-properties-panel__label"
            >
              Next Stop
            </div>
            <div
              className="m-vehicle-properties-panel__value"
            >
              Stop Name
            </div>
          </div>
          <div
            className="m-vehicle-properties-panel__map"
          >
            <div
              className="m-vehicle-map-state m-vehicle-map-state--auto-centering"
            />
            <div
              className="m-vehicle-map"
              id="id-vehicle-map"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="m-properties-panel__modal-overlay"
    onClick={[Function]}
  />,
]
`;
